contracts/actions/mcd/McdRepayComposite.sol

```diff
@@ -0,0 +1,268 @@
+    function receiveFlashLoan(
+        address[] memory _tokens,
+        uint256[] memory _amounts,
+        uint256[] memory _feeAmounts,
+        bytes memory _userData
+    ) external override nonReentrant {
+        require(msg.sender == VAULT_ADDR, "Untrusted lender");
+        (RepayParams memory repayParams, address proxy) = abi.decode(_userData, (RepayParams, address));
+
+        _repay(proxy, repayParams);
+
+        uint256 flPaybackAmount = _amounts[0] + _feeAmounts[0];
+        _tokens[0].withdrawTokens(address(VAULT_ADDR), flPaybackAmount);
+    }
+
+    function _repay(address _proxy, RepayParams memory _repayParams) internal {
+        uint256 collateral = getAllColl(IManager(_repayParams.mcdManager), _repayParams.joinAddr, _repayParams.vaultId);
+        _repayParams.repayAmount = _repayParams.repayAmount > collateral ? collateral : _repayParams.repayAmount;
+
+        // Sell flashloaned collateral asset for debt asset
+        assert(_repayParams.exchangeData.srcAmount == _repayParams.repayAmount); // CONSIDER
+        assert(_repayParams.exchangeData.destAddr == DAI_ADDR); // CONSIDER
+        (uint256 exchangedAmount, ) = _dfsSell(_repayParams.exchangeData, address(this), address(this), false);
+
+        (address urn, bytes32 ilk) = getUrnAndIlk(_repayParams.mcdManager, _repayParams.vaultId);
+        uint256 paybackAmount = exchangedAmount;
+        // if paybackAmount is higher than current debt, repay all debt and send remaining dai to proxy
+        {
+            uint256 debt = getAllDebt(address(vat), urn, urn, ilk);
+            if (exchangedAmount > debt) {
+                DAI_ADDR.withdrawTokens(IDSProxy(_proxy).owner(), exchangedAmount - debt);
+                paybackAmount = debt;
+            }
+        }
+        DAI_ADDR.approveToken(DAI_JOIN_ADDR, paybackAmount);
+
+        {
+            address joinToken = _repayParams.exchangeData.srcAddr;
+            uint256 collateralAssetBalanceDelta = joinToken.getBalance(address(this));
+
+            if (_repayParams.mcdManager == CROPPER) {
+                _cropperRepay(
+                    _repayParams.vaultId,
+                    urn,
+                    ilk,
+                    _repayParams.joinAddr,
+                    paybackAmount,
+                    _repayParams.repayAmount
+                );
+            } else {
+                _mcdManagerRepay(
+                    _repayParams.mcdManager,
+                    _repayParams.vaultId,
+                    urn,
+                    ilk,
+                    _repayParams.joinAddr,
+                    paybackAmount,
+                    _repayParams.repayAmount
+                );
+            }
+            collateralAssetBalanceDelta = joinToken.getBalance(address(this)) - collateralAssetBalanceDelta;
+            require(collateralAssetBalanceDelta == _repayParams.repayAmount);
+        }
+
+        logger.logRecipeEvent("McdRepayComposite");
```

Comment:
unsure if these two should be split, one in action direct one in recipe execution